File: solo-leveling-ui-service/src/main/java/com/sleepkqq/sololeveling/ui/aop/LoggingAspect.java
package com.sleepkqq.sololeveling.ui.aop;

import com.sleepkqq.sololeveling.ui.model.UserData;
import com.sleepkqq.sololeveling.ui.service.TgAuthService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.util.StopWatch;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.stereotype.Component;

@Slf4j
@Aspect
@Component
@RequiredArgsConstructor
public class LoggingAspect {

  private final TgAuthService tgAuthService;

  @Pointcut("within(com.sleepkqq.sololeveling.ui.api..*)")
  public void apiMethods() {}

  @Around("apiMethods()")
  public Object logExecutionTime(ProceedingJoinPoint joinPoint) throws Throwable {
    var username = tgAuthService.findCurrentUser().map(UserData::getUsername).orElse("anon");
    var stopWatch = new StopWatch();

    stopWatch.start();
    var result = joinPoint.proceed();
    stopWatch.stop();

    log.info(
        ">> api '{}' executed by '{}' in {} ms",
        joinPoint.getSignature().getName(),
        username,
        stopWatch.getTotalTimeMillis()
    );

    return result;
  }
}

File: solo-leveling-gateway-service/src/main/kotlin/com.sleepkqq.sololeveling/gateway/aop/LoggingAspect.kt
package com.sleepkqq.sololeveling.gateway.aop

import com.sleepkqq.sololeveling.config.interceptor.UserContextHolder
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.util.StopWatch

@Aspect
@Component
class LoggingAspect {

	private val log = LoggerFactory.getLogger(javaClass)

	@Pointcut("within(com.sleepkqq.sololeveling.gateway.controller..*)")
	fun controllersMethods() {
	}

	@Pointcut("within(com.sleepkqq.sololeveling.gateway.grpc.client..*)")
	fun grpcClientsMethods() {
	}

	@Around("controllersMethods()")
	@Throws(Throwable::class)
	fun controllersMethodsLogger(joinPoint: ProceedingJoinPoint): Any =
		logExecutionTime(joinPoint, ExecutionType.REST_API)

	@Around("grpcClientsMethods()")
	@Throws(Throwable::class)
	fun grpcClientsMethodsLogger(joinPoint: ProceedingJoinPoint): Any =
		logExecutionTime(joinPoint, ExecutionType.GRPC_CALL)

	private fun logExecutionTime(
		joinPoint: ProceedingJoinPoint,
		executionType: ExecutionType
	): Any {
		val stopWatch = StopWatch()
		val userId = UserContextHolder.getUserId()?.toString() ?: "unauthenticated"

		stopWatch.start()
		val result = try {
			joinPoint.proceed()

		} finally {
			if (stopWatch.isRunning) {
				stopWatch.stop()
			}
		}

		log.info(
			"{} {} '{}' executed by '{}' in {} ms",
			executionType.operationSymbol.value,
			executionType.value,
			joinPoint.signature.name,
			userId,
			stopWatch.totalTimeMillis
		)

		return result
	}
}
